### DISCUSSION TITLE
LLCC68 Initialization

### DISCUSSION THREAD

**Discussion by Eugene-long** [jgromes/RadioLib#1315] [Q&A] [answered]
Previously purchased E220 (PLCC68) boards work without problems. I bought 2 new E220 boards (LLCC68). There was a problem with initialization. The problem with setting the spread factor:
12:10:23.882 -> [LLCC68] Initialization 1.. The selected expansion coefficient is not valid for this module!

I chose a value from 7 to 12. The remaining parameters are set successfully. What could be the problem?

The part of sketch: 
// set spreading factor to 7
  if (radio1.setSpreadingFactor(7) == RADIOLIB_ERR_INVALID_SPREADING_FACTOR) {
    Serial.println(F("Selected spreading factor is invalid for this module!"));
    while (true);
  }

**Comment by Eugene-long** [comment, marked as answer]
E220 (PLCC68) = E220 (LLCC68)

**Comment by jgromes** [comment]
> 12:10:23.882 -> [LLCC68] Initialization 1.. The selected expansion coefficient is not valid for this module!

I don't know what an "expansion coefficient" is, but assuming you're talking about spreading factor, LLCC68 is more limited in its configuration compared to SX126x, it only supports 125, 250 and 500 kHz bandwidth, and a limited selection of spreading factors (SF5 - 9 for BW 125, SF5 - 10 for BW 250 and SF5-11 for BW 500). My guess is that you have an invalid combination of spreading factor and bandwidth, though the library should be able to prevent that.

**Reply by Eugene-long** [reply]
Hi Jan, thank you for reply. I try to set up BW=125 and SF=7. Does the Library check which SF and BW values are actually written in the register?

73 Eugene

**Reply by jgromes** [reply]
It verifies whether the combination is valid, but on SX126x, there is actually no way of verifying whether the configuration is correct by reading the SPI register (as the values is sent via a SPI command, not an SPI register write). See here: https://github.com/jgromes/RadioLib/blob/fe52311371dbde753555b6afb9472cd5db900174/src/modules/LLCC68/LLCC68.cpp#L38-L54

**Reply by Eugene-long** [reply]
Thank you

**Comment by Eugene-long** [comment]
Hi Jan, thank you for reply. I try to set up BW=125 and SF=7. These values are correct. Why I see this message?
12:10:23.882 -> [LLCC68] Initialization 1.. Selected spreading factor is invalid for this module! 

May be this module E220 is bad? But the second one give the same message.

73! Eugene

**Comment by Eugene-long** [comment]
The part of scketch:

Serial.print(F("[LLCC68] Initializing 2.. "));
    int state = radio1.begin();
  
  // carrier frequency:           433.0 MHz
  // bandwidth:                   125.0 kHz
  // spreading factor:            7
  // coding rate:                 5
  // sync word:                   0x12 (`0x12`как в приемнике на SX1278)
  // output power:                20 dBm
  // preamble length:             8 symbols
  state = radio1.begin(433.0, 125.0, 7, 5, 0x12, 20, 8);
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while (true);

  }

Message:
12:23:26.601 -> [LLCC68] Initializing 1.. [LLCC68] Initializing 2.. failed, code -2

Could you please explain what "code -2" means?

**Reply by jgromes** [reply]
> I try to set up BW=125 and SF=7. These values are correct. Why I see this message?

I don't know, you have not shared your code. I was not able to replicate your problem with BW125 SF7, so I'm fairly confident the issue is nto in the library.

> Could you please explain what "code -2" means?

See the full list of status codes [here](https://jgromes.github.io/RadioLib/group__status__codes.html) or an online decoder [here](https://radiolib-org.github.io/status_decoder/decode.html). Regarding the code snippet you have shared, why are you calling the begin method twice?

**Reply by Eugene-long** [reply]
Hi, Jan,
the problem is solved. The E220-400M22S module with serial number S3202935S40128 is based on the LLCC68 chip indeed. The same modules with serial numbers S4202060S05196, S4202060S05184 based on the SX1262 chip. I think this is a manufacturer's problem.
Thank you."

Eugene

**Reply by Eugene-long** [reply]
I've spent some time researching this problem. You're right. The issue is not in the library. The E220-400M22S module with serial number S3202935S40128 works as LLCC68 chip, and the same labeled (E220-400M22S)  module with serial number S4202060S05196  works as SX1262 chip. These chips have similar parameters, but there are differences. 
Thanks for the assistance.
Eugene
